File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/AnimatedListMenuView.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "AnimatedListMenuView.h"
#import "AnimatedListMenuView+Internal.h"
#import "Wire-Swift.h"

@implementation MenuDotView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.layer.borderColor = [UIColor whiteColor].CGColor;
        self.layer.borderWidth = 2.0;
    }
    return self;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    self.layer.cornerRadius = self.bounds.size.width / 2;
}

@end


@interface AnimatedListMenuView ()

@end

@implementation AnimatedListMenuView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.leftDotView = [[MenuDotView alloc] init];
        [self addSubview:self.leftDotView];
        
        self.centerDotView = [[MenuDotView alloc] init];
        [self addSubview:self.centerDotView];
        
        self.rightDotView = [[MenuDotView alloc] init];
        [self addSubview:self.rightDotView];
    }
    return self;
}

- (void)setProgress:(CGFloat)progress
{
    if (_progress == progress) {
        return;
    }
    _progress = MIN(1, MAX(0, progress));
}

- (void)setProgress:(CGFloat)progress animated:(BOOL)animated
{
    [self setProgress:progress];
    
    self.centerToRightDistanceConstraint.constant = [self centerToRightDistanceForProgress:self.progress];
    self.leftToCenterDistanceConstraint.constant = [self leftToCenterDistanceForProgress:self.progress];

    if (animated) {
        [self setNeedsUpdateConstraints];
        
        [UIView animateWithDuration:0.25f animations:^{
            [self layoutIfNeeded];
        }];
    }
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/ListContent/AnimatedListMenuView.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

// MARK: - MenuDotView

private final class MenuDotView: UIView {
    // MARK: Lifecycle

    init() {
        super.init(frame: .zero)

        layer.borderColor = UIColor.white.cgColor
        layer.borderWidth = 2
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: Internal

    override func layoutSubviews() {
        super.layoutSubviews()

        layer.cornerRadius = bounds.size.width / 2
    }
}

// MARK: - AnimatedListMenuView

final class AnimatedListMenuView: UIView {
    // MARK: Lifecycle

    convenience init() {
        self.init(frame: .zero)
    }

    override init(frame: CGRect) {
        super.init(frame: frame)

        for item in [leftDotView, centerDotView, rightDotView] {
            addSubview(item)
        }
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: Internal

    /// Animation progress. Value from 0 to 1.0
    var progress: CGFloat = 0 {
        didSet {
            guard !(0 ... 1 ~= progress) else { return }
            progress = min(1, max(0, progress))
        }
    }

    func setProgress(_ progress: CGFloat, animated: Bool) {
        self.progress = progress

        centerToRightDistanceConstraint?.constant = centerToRightDistance(forProgress: self.progress)
        leftToCenterDistanceConstraint?.constant = leftToCenterDistance(forProgress: self.progress)

        if animated {
            setNeedsUpdateConstraints()

            UIView.animate(withDuration: 0.25, animations: {
                self.layoutIfNeeded()
            })
        }
    }

    override func updateConstraints() {
        if initialConstraintsCreated {
            super.updateConstraints()
            return
        }

        let dotWidth: CGFloat = 4

        let dotViews = [leftDotView, centerDotView, rightDotView]

        dotViews.forEach { $0.translatesAutoresizingMaskIntoConstraints = false }
        translatesAutoresizingMaskIntoConstraints = false

        centerToRightDistanceConstraint = centerDotView.rightAnchor.constraint(
            equalTo: rightDotView.leftAnchor,
            constant: centerToRightDistance(forProgress: progress)
        )

        leftToCenterDistanceConstraint = leftDotView.rightAnchor.constraint(
            equalTo: centerDotView.leftAnchor,
            constant: leftToCenterDistance(forProgress: progress)
        )

        let leftDotLeftConstraint = leftDotView.leftAnchor.constraint(equalTo: leftAnchor)

        for dotView in dotViews {
            dotView.widthAnchor.constraint(equalToConstant: dotWidth).isActive = true
            dotView.heightAnchor.constraint(equalToConstant: dotWidth).isActive = true
        }

        let subviewConstraints: [NSLayoutConstraint] = [
            leftDotView.centerYAnchor.constraint(equalTo: centerYAnchor),
            centerDotView.centerYAnchor.constraint(equalTo: centerYAnchor),
            rightDotView.centerYAnchor.constraint(equalTo: centerYAnchor),

            rightDotView.rightAnchor.constraint(equalTo: rightAnchor, constant: -8),
            leftDotLeftConstraint,

            centerToRightDistanceConstraint!,
            leftToCenterDistanceConstraint!,
        ]
        NSLayoutConstraint.activate(subviewConstraints)

        initialConstraintsCreated = true

        super.updateConstraints()
    }

    func centerToRightDistance(forProgress progress: CGFloat) -> CGFloat {
        -(4 + (10 * (1 - progress)))
    }

    func leftToCenterDistance(forProgress progress: CGFloat) -> CGFloat {
        -(4 + (20 * (1 - progress)))
    }

    // MARK: Private

    private let leftDotView = MenuDotView()
    private let centerDotView = MenuDotView()
    private let rightDotView = MenuDotView()

    private var initialConstraintsCreated = false
    private var centerToRightDistanceConstraint: NSLayoutConstraint?
    private var leftToCenterDistanceConstraint: NSLayoutConstraint?
}
